File: solutions/cpp/3.longest-substring-without-repeating-characters.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() == 0) return 0;
        int res = 0, n = s.length();
        unordered_map<int, int> m;
        for (int left = -1, right = 0; right < n; ++right) {
            if (m.count(s[right]) && m[s[right]] > left) {
                left = m[s[right]];
            }
            m[s[right]] = right;
            res = max(res, right - left);
        }
        return res;
    }
};

int main() {
    Solution soln;
    string s = "abcabcbb";
    int res = soln.lengthOfLongestSubstring(s);
    cout << "string: " << s << endl;
    cout << "length of longest substring: " << res << endl;
    return 0;
}

File: solutions/python/3.longest-substring-without-repeating-characters.py
# 3. Longest Substring Without Repeating Characters


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_map = dict()
        ansidx = 0
        longest = 0
        i = 0
        for c in s:
            if c in char_map:
                if not char_map[c] < ansidx:
                    ansidx = char_map[c] + 1
            char_map[c] = i
            sublen = i - ansidx + 1
            if sublen > longest:
                longest = sublen
            i = i + 1
        return longest


def test():
    s = Solution()

    string = 'abcabcbb'
    output = s.lengthOfLongestSubstring(string)
    print(f'{string=}\n{output=}\n')

    string = 'bbbbb'
    output = s.lengthOfLongestSubstring(string)
    print(f'{string=}\n{output=}\n')

    string = 'pwwkew'
    output = s.lengthOfLongestSubstring(string)
    print(f'{string=}\n{output=}\n')

    string = 'dvdf'
    output = s.lengthOfLongestSubstring(string)
    print(f'{string=}\n{output=}')


if __name__ == '__main__':
    test()
